packages/hash/api/src/graphql/resolvers/pages/setParentPage.ts

```diff
@@ -1,87 +1,121 @@
-        ? `'${parentPageId}'`
-        : `'${pageId}'${
-            parentPageEntity?.entityId
-              ? ""
-              : ` nor parent page with entityId = '${parentPageId}'`
-          }`;
-      throw new ApolloError(
-        `Could not find entity with entityId = ${notFoundId}.`,
-        "NOT_FOUND",
-      );
-    }
-
-    const pages = await Page.getAccountPagesWithParents(dataSources.db, {
+> = async (
+  _,
+  { accountId, pageId, parentPageId },
+  { dataSources: { db }, user },
+) => {
+  return await db.transaction(async (client) => {
+    const pageEntity = await Page.getAccountPageWithParents(client, {
       accountId,
-      systemTypeName: "Page",
+      entityId: pageId,
     });
 
-    // Check for cyclic dependencies.
-    try {
-      const directedPages = pages
-        .filter((page) => page.parentEntityId != null)
-        .map(
-          (page) => [page.entityId, page.parentEntityId] as [string, string],
-        );
-      // add the new parent relation to the existing relations
-      directedPages.push([pageId, parentPageId]);
-      topologicalSort(directedPages);
-    } catch (_error) {
+    if (!pageEntity) {
       throw new ApolloError(
-        `Could not set '${parentPageId}' as parent to '${pageId}' as this would create a cyclic dependency.`,
-        "CYCLIC_TREE",
+        `Could not find page entity with entityId = ${pageId} on account with id = ${accountId}.`,
+        "NOT_FOUND",
       );
     }
-
-    if (pageEntity.parentEntityId) {
-      const link = await pageEntity.getOutgoingLinkByEntityId(client, {
-        destinationEntityId: pageEntity.parentEntityId,
-      });
-      if (!link) {
-        throw new Error(
-          `Could not find existing link to parent page = '${pageEntity.parentEntityId}'.`,
-        );
-      }
-
-      await pageEntity.deleteOutgoingLink(client, {
-        linkId: link.linkId,
-        deletedByAccountId: user.accountId,
-      });
+    // Either replace a parent page link, or delete it depending on whether or not `parentPageId` is set.
+    if (parentPageId) {
+      await replaceParentPage(db, user, accountId, pageEntity, parentPageId);
+    } else {
+      await deleteParentPage(db, user, pageEntity);
     }
 
-    // @todo: lock entity to prevent potential race-condition when updating entity's properties
-    await pageEntity.createOutgoingLink(client, {
-      createdByAccountId: user.accountId,
-      stringifiedPath: "$.parent",
-      destination: parentPageEntity,
+    // @todo: allow to refetch account page with parents
+    // Refetched page exists here, but it could fail (return null) if the entity were to be removed before this point.
+    // A Page model method to refetch would be ideal here.
+    const refetchedPage = await Page.getAccountPageWithParents(client, {
+      accountId,
+      entityId: pageId,
     });
 
-    return pageEntity.toGQLUnknownEntity();
+    return refetchedPage!.toGQLPageEntity();
```

Comment:
Note: you won't have to do this if you implement the `parentPage` field resolver